### ISSUE TITLE
Changing API?

### ISSUE THREAD

**Issue by kodonnell** [adamcharnock/mate3#4]
Just wondering about some possible changes, as below. Note that it wouldn't make sense to make all of these - some remove the need for others.

### 1. Parser -> Device

Just thinking that `ChargeControllerParser` confuses me - it's somewhat internal logic, whereas as a user I want to think about `ChargeController` as a device. So instead of

```python
from mate3.parsers import ChargeControllerParser
values = client.get_values(
    ChargeControllerParser.battery_current, 
    ChargeControllerParser.battery_voltage
)
```

I'd be less confused with

```python
from mate3.devices import ChargeController
values = client.get_values(
    ChargeController.battery_current, 
    ChargeController.battery_voltage
)
```

### 2. Remove use of enums/lookups and use objects for devices

For example, instead of 

```python
client.get_values(ChargeControllerParser.battery_current)
```

What do you think about?

```python
client.charge_controllers[0].battery_current
```

i.e. 
- automatically have the client read the charge controller into an object, and
- use the native `__getattr__` instead of reimplementing this as `get_values`. And we could ensure the appropriate scaling is applied before returning to the user too, actually.

Similarly, it makes writing easier/cleaner. Instead of 

```python
client.set_value(
    field=ChargeControllerConfigurationParser.absorb_volts,
    value=330,
    port=3
)
```

we could have

```python
client.charge_controllers[0].absorb_volts = 33 # NB: internally we scale it appropriately before sending to modbus
client.write() # (or we could have an 'autocommit' functionality)
)
```

This is similar to how [pySunSpec](https://pysunspec.readthedocs.io/en/latest/pysunspec.html#interacting-with-a-sunspec-device) do it.

### Remove modbus references (at least at high level)

For example, allow the user to interact without knowing what blocks are. If I include the changes from 2. and this, the main example from the readme could look like this:

```python
from mate3 import mate3_connection

host = '192.168.0.123'
port = 502

with mate3_connection(host, port) as client:
    client.read()
    print(client.charge_controllers[0].battery_current)
    print(client.charge_controllers[0].battery_voltage)
    # Writing data
    client.charge_controllers[0].absorb_volts = 33
    client.write()
```

(Note that I've merged the `ChargeController` and `ChargeControllerConfiguration`.)

Thoughts? I'm just thinking that for users like me, who just want to read (and maybe write) values, we try to make it as easy as possible. There are some minor things to sort out, but if you like the general idea, I can have a crack at a PR?

**Comment by adamcharnock** [comment]
Hi @kodonnell,

I've had a scan over this but not read it in detail. However, this all sounds like exactly the kind of thing I would do, so I'm broadly onboard.

Can you also see my work-in-progress branch `issues/3`. I think I have made some progress on some major refactoring, but I cannot remember precisely what I did or how far I got. Perhaps take a look at that and see what you think.

I'm definitely keen to have help with this project though, and your suggestions sound excellent.

**Comment by adamcharnock** [comment]
To add to the above: Take a look at `issues/3` and then leave another comment with your thoughts. I'll read into your notes above in more detail over the coming days. If you don't hear from me in a few days then definitely leave a comment here to remind me.

**Comment by kodonnell** [comment]
> To add to the above: Take a look at issues/3

Will do when you push it = ) If it doesn't look like too much work I may just whip up a quick POC too.

**Comment by adamcharnock** [comment]
**Head desk** 🙄 

Pushed!

**Comment by kodonnell** [comment]
Thanks. Just checking - I can only see [one commit](https://github.com/adamcharnock/mate3/commit/66a452536160001a908499fa42bd66989c56bfbb) which is just renaming `parser` to `definition`, so given what you've said above, maybe some commits are missing?

**Comment by adamcharnock** [comment]
Ok, it looks like I perhaps didn't get very far (or I may have rolled back some of my local commits because it didn't work out). In which case, definitely feel free to take a stab at this.

**Comment by kodonnell** [comment]
Cool. Just thinking this through now, and I'm wondering about performance - and you may have more experience with this. Doing a `client.read()` which updates all the values for every device is simple and clean ...

```python
client.read()
client.charge_controllers[0].battery_current
```

... but I'm worried about the performance impact of doing that. Presumably, the below will still be faster:

```python
client.get_values(ChargeControllerParser.battery_current)
```

As it's only getting a single field ... right?

I guess the question becomes what is 'faster' - if it can read everything in 10ms, then who cares, but if it's 1 second (vs 10ms for a single field) then it's worth thinking about. Do you have any idea? On my system, running the `mate3` command (on a mate3 not a mate3s) takes about 1 second (piping to /dev/null) but I don't know how much of that is connection latency. If I use the python API example, then reading all blocks is about 90ms and similarly for a single block, while reading a single field is more like 30ms. And then it bricked ... so maybe the mate3 doesn't like repeated (sub second) "select *" over modbus?

That all said, I doubt most people will be monitoring values faster that 1s, in which case convenience may win. For the crazy people who want it that fast (or maybe faster in my case, if possible - I want to control variable loads dynamically), then it's probably relatively easy to add the equivalent of `only_fields` into `client.read()` etc. On this note, I was already thinking about having `time_last_read` and `is_stale` etc. on fields anyway, so it could be valid to e.g. do `client.read()` once every minute, and specific fields `client.read(only_fields=...)` e.g. every second - at any point in time, you'd have access to all fields (assuming you've done at least one `client.read`), though some of the values might just be a minute old (which, presumably, you don't really care about - e.g. battery SOC doesn't change that fast). But you can tell how old they are, etc.

Anyway, what are your thoughts? I think either way it's still nice to do `client.charge_controllers[0].battery_current` - the question is really just the optimal way to update it. Is it from a `client.read` (potentially limited to only some fields) ... or we could even hit modbus with every `__getattr__` ... but that sounds inefficient for large numbers of fields, as per your reading range work so far.

**Comment by adamcharnock** [comment]
> That all said, I doubt most people will be monitoring values faster that 1s, in which case convenience may win. For the crazy people who want it that fast (or maybe faster in my case, if possible - I want to control variable loads dynamically), then it's probably relatively easy to add the equivalent of only_fields into client.read() etc.

I think this is a sensible approach. Read all data, the if anyone needs to read only single fields then we/they can add that feature. Also, I think the values on the mate3 don't refresh that fast anyway (maybe once every 1-3seconds).

I also think the timings you measured seem representative of what I experienced, but I don't have exact numbers to hand.

FYI – The first iteration of this library read all fields in separate read commands, which took in the region of 1 second.

> And then it bricked

I see this happen when connections to the mate3 are not nicely cleaned up. Also, the mate3 can only handle one connection at once. I therefore create & destroy a fresh connection for every polling loop.

> On this note, I was already thinking about having `time_last_read` and `is_stale` etc. on fields anyway, so it could be valid to e.g. `do client.read()` once every minute, and specific fields `client.read(only_fields=...)`

Sounds cool, but perhaps something to in a second pass, once the initial refactoring is done and merged.

> the question is really just the optimal way to update it

Personally I would keep it simple for now, and just allow the user to manually trigger a refresh whenever they wish. Once that is working, and we are happy with it, we can start to think about fancier stuff 😄 

---

As an added note: I have a spreadsheet from OutBack which gives an updated list of registers for the latest firmware. I say this because:

* This will allow us the ditch the PDF parsing code
* The code generation will need to be maintained (in fact, don't even edit the generated code directly. Instead, modify the code generation routines, then rerun the code generation)

**Comment by kodonnell** [comment]
> FYI – The first iteration of this library read all fields in separate read commands, which took in the region of 1 second.

Interesting - why the change to what it is now? I'm still relatively unfamiliar with the code, so don't want to make the wrong call out of naivete.

> Sounds cool, but perhaps something to in a second pass, once the initial refactoring is done and merged.

Nicely put = ) Yes, my thoughts for potential PRs (which may not quite work in reality as they may depend on each other) are:

1. Continue your refactoring. Parsers -> definitions, etc. and maybe combining the multiple charge controller blocks into one, etc.
2. Move to returning objects which have the properties on them.
3. Remove most of the modbus'y stuff and go with simple `client.read` etc.
4. Add in `client.read(fields=...)`
5. Add `client.write` after setting attributes on the object etc. Make it smart and only write changed values, etc.

Later:

1. Auto-scale the attributes appropriately.
2. Add `time_last_updated` and `is_stale` niceness.
3. Autocommit on write.

> As an added note: I have a spreadsheet from OutBack which gives an updated list of registers for the latest firmware. 

Oh, that sounds bad - breaking changes, or just new registers? I just updated my mate3 to the latest, and this code seems to work fine (though I haven't checked everything). Anyway, that can be a separate PR?

**Comment by kodonnell** [comment]
> As an added note: I have a spreadsheet from OutBack which gives an updated list of registers for the latest firmware. I say this because:

Oh - can you post it? I was needing to update that anyway, and it turns out the PDF extraction is a lot slower than I thought, so a spreadsheet would be nice. Also - do you mind if I commit it in if it's not too big? Since we're using it as code generation, I prefer to guarantee everyone has the same version etc.

**Comment by adamcharnock** [comment]
I *think* it is ok for me to post this here... 
[OutBack Power SunSpec Map.xlsx](https://github.com/adamcharnock/mate3/files/4643402/OutBack.Power.SunSpec.Map.xlsx)

**Comment by adamcharnock** [comment]
> do you mind if I commit it in if it's not too big?

I committed in the interim CSV files, which is what the code generation builds from. There is a separate process which converts PDF->CSV. Given the (lightweight) CSVs were present, and sufficient for code generation, I did not include the PDF in git.

**Comment by kodonnell** [comment]
Closed by #6
